src/main/java/co/cask/hydrator/plugin/source/batch/SalesforceBatchSource.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.source.batch;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Macro;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.batch.Input;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.api.batch.BatchSourceContext;
+import co.cask.hydrator.common.LineageRecorder;
+import co.cask.hydrator.plugin.BaseSalesforceConfig;
+import co.cask.hydrator.salesforce.SalesforceSchemaUtil;
+import co.cask.hydrator.salesforce.authenticator.AuthenticatorCredentials;
+import co.cask.hydrator.salesforce.parser.SalesforceQueryParser;
+import com.sforce.ws.ConnectionException;
+import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.LocalDate;
+import java.util.stream.Collectors;
+import javax.ws.rs.Path;
+
+
+/**
+ * Plugin returns records from Salesforce using provided by user SOQL query.
+ * Salesforce bulk API is used to run SOQL query. Bulk api returns data in batches.
+ * Every batch is processed as a separate split by mapreduce.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(SalesforceBatchSource.NAME)
+@Description("Plugin to read data from Salesforce in batches.")
+public class SalesforceBatchSource extends BatchSource<String, String, StructuredRecord> {
+  static final String NAME = "SalesforceBatchSource";
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceBatchSource.class);
+
+  private final Config config;
+  private Schema schema;
+
+  SalesforceBatchSource(Config config) throws ConnectionException {
+    this.config = config;
+  }
+
+  static final class Config extends BaseSalesforceConfig {
+    @Description("The SOQL query to retrieve results for")
+    @Macro
+    private final String query;
+
+    Config(Configuration conf) {
+        super(null, conf.get(SalesforceConstants.CLIENT_ID), conf.get(SalesforceConstants.CLIENT_SECRET),
```

Comment:
Why reference name is null?

Reply:
This config is re-generated in mapreduce via this constructror and than used for authentication etc. We don't have and don't need and don't have referenceName there